File: 01a/parser_python/DataParser.py
import json
import yaml
import xml.etree.ElementTree as ET
import csv

def read_txt(file_path):
    with open(file_path, "r", encoding="utf-8") as file:
        return file.read() # her sker integration: læser og integrerer indholdet af tekstfilen i systemet

def read_json(file_path):
    with open(file_path, "r", encoding="utf-8") as file:
        return json.load(file) # her sker integration: læser og parser JSON-filen, så data kan bruges i Python

def read_yaml(file_path):
    with open(file_path, "r", encoding="utf-8") as file:
        return yaml.safe_load(file) # her sker integration: læser og parser YAML-filen

def read_xml(file_path):
    tree = ET.parse(file_path) # her sker integration: læser og parser XML-filen fra filsystemet
    root = tree.getroot()
    return {child.tag: child.text for child in root}

def read_csv(file_path):
    with open(file_path, "r", encoding="utf-8") as file:
        reader = csv.DictReader(file)
        return [row for row in reader] # her sker integration: læser og parser CSV-filen til Python-objekter

if __name__ == "__main__":
    print("TXT Data:", read_txt("../data_files/data.txt"))
    print("JSON Data:", read_json("../data_files/data.json"))
    print("YAML Data:", read_yaml("../data_files/data.yaml"))
    print("XML Data:", read_xml("../data_files/data.xml"))
    print("CSV Data:", read_csv("../data_files/data.csv"))

File: 01a/parser_node/DataParser.js
const fs = require("fs");
const yaml = require("js-yaml");
const xml2js = require("xml2js");
const csv = require("csv-parser");
const path = require("path"); // Hjælper med korrekte filstier

// Funktion til at generere den rigtige sti til filerne
const getFilePath = (filename) => path.join(__dirname, "../data_files", filename); // integration sker her: dynamisk filstihåndtering, så alle parser-funktioner kan finde de rigtige datafiler

// Læs TXT-fil
const readTxt = () => fs.readFileSync(getFilePath("data.txt"), "utf8"); // integration sker her: læser og integrerer indholdet af tekstfilen i systemet

// Læs JSON-fil
const readJson = () => JSON.parse(fs.readFileSync(getFilePath("data.json"), "utf8")); // integration sker her: læser og parser JSON-filen, så data kan bruges i JavaScript

// Læs YAML-fil
const readYaml = () => yaml.load(fs.readFileSync(getFilePath("data.yaml"), "utf8")); // integration sker her: læser og parser YAML-filen

// Læs XML-fil (async fordi det bruger en parser)
const readXml = async () => {
    const data = fs.readFileSync(getFilePath("data.xml"), "utf8"); // integration sker her: læser XML-filen fra filsystemet
    return new Promise((resolve, reject) => {
        xml2js.parseString(data, (err, result) => {
            if (err) reject(err);
            else resolve(result);
        });
    });
};

// Læs CSV-fil (async)
const readCsv = async () => {
    return new Promise((resolve, reject) => {
        const results = [];
        fs.createReadStream(getFilePath("data.csv")) // integration sker her: læser CSV-filen fra filsystemet
            .pipe(csv()) // integration sker her: parser CSV-data til JavaScript-objekter
            .on("data", (data) => results.push(data))
            .on("end", () => resolve(results))
            .on("error", (err) => reject(err));
    });
};

// MAIN: Kør parseren
const main = async () => {
    try {
        console.log("TXT Data:", readTxt());
        console.log("JSON Data:", readJson());
        console.log("YAML Data:", readYaml());
        console.log("XML Data:", await readXml());
        console.log("CSV Data:", await readCsv());
    } catch (err) {
        console.error("Fejl:", err.message);
    }
};

main();
